File: MelonLoader/CommandLine.cs
﻿using System;
using System.Collections.Generic;
using System.IO;

namespace MelonLoader
{
    public static class MelonCommandLine
    {
        private class ArgInfo
        {
            internal string CommandLine;
            internal delegate void ParseCommandLineArgDelegate(string value);
            internal ParseCommandLineArgDelegate ParseCommandLineArg;

            internal string Config_Name;
            internal string Config_Category;
            internal delegate void ParseConfigArgDelegate(ArgInfo argInfo);
            internal ParseConfigArgDelegate ParseConfigArg;
        }
        private static readonly List<ArgInfo> ArgInfoTbl = new List<ArgInfo>();
        private static readonly IniFile iniFile = null;

        static MelonCommandLine()
        {
            iniFile = new IniFile(Path.Combine(Path.Combine(MelonUtils.GameDirectory, "MelonLoader"), "LaunchOptions.ini"));

            string Config_Category_AnalyticsBlocker = "AnalyticsBlocker";
            string Config_Category_AssemblyGenerator = "AssemblyGenerator";
            string Config_Category_Console = "Console";
            string Config_Category_Core = "Core";
            string Config_Category_Logger = "Logger";

            #region AnalyticsBlocker
            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.dab",
                ParseCommandLineArg = (string value) => AnalyticsBlocker.ShouldDAB = true,

                Config_Name = "ShouldDAB",
                Config_Category = Config_Category_AnalyticsBlocker,
                ParseConfigArg = (ArgInfo argInfo) => AnalyticsBlocker.ShouldDAB = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, AnalyticsBlocker.ShouldDAB, true)
            });
            #endregion

            #region AssemblyGenerator
            /*
            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.agfregenerate",
                ParseCommandLineArg = (string value) => AssemblyGenerator.ForceRegeneration = true,

                //Config_Name = "ForceRegenerate",
                //Config_Category = Config_Category_AssemblyGenerator,
                //ParseConfigArg = (ArgInfo argInfo) => AssemblyGenerator.ForceRegeneration = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, AssemblyGenerator.ForceRegeneration, true)
            });
            */

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.agfvdumper",
                ParseCommandLineArg = (string value) => AssemblyGenerator.ForceVersion_Dumper = value,

                Config_Name = "ForceVersion_Dumper",
                Config_Category = Config_Category_AssemblyGenerator,
                ParseConfigArg = (ArgInfo argInfo) => AssemblyGenerator.ForceVersion_Dumper = iniFile.GetString(argInfo.Config_Category, argInfo.Config_Name, AssemblyGenerator.ForceVersion_Dumper, true)
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.agfvunhollower",
                ParseCommandLineArg = (string value) => AssemblyGenerator.ForceVersion_Il2CppAssemblyUnhollower = value,

                Config_Name = "ForceVersion_Il2CppAssemblyUnhollower",
                Config_Category = Config_Category_AssemblyGenerator,
                ParseConfigArg = (ArgInfo argInfo) => AssemblyGenerator.ForceVersion_Il2CppAssemblyUnhollower = iniFile.GetString(argInfo.Config_Category, argInfo.Config_Name, AssemblyGenerator.ForceVersion_Il2CppAssemblyUnhollower, true)
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.agfvunity",
                ParseCommandLineArg = (string value) => AssemblyGenerator.ForceVersion_UnityDependencies = value,

                Config_Name = "ForceVersion_UnityDependencies",
                Config_Category = Config_Category_AssemblyGenerator,
                ParseConfigArg = (ArgInfo argInfo) => AssemblyGenerator.ForceVersion_UnityDependencies = iniFile.GetString(argInfo.Config_Category, argInfo.Config_Name, AssemblyGenerator.ForceVersion_UnityDependencies, true)
            });
            #endregion

            #region Console
#if !DEBUG
            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.hideconsole",
                ParseCommandLineArg = (string value) => Console.Enabled = false,

                Config_Name = "Enabled",
                Config_Category = Config_Category_Console,
                ParseConfigArg = (ArgInfo argInfo) => Console.Enabled = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, Console.Enabled, true)
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.hidewarnings",
                ParseCommandLineArg = (string value) => Console.HideWarnings = true,

                Config_Name = "HideWarnings",
                Config_Category = Config_Category_Console,
                ParseConfigArg = (ArgInfo argInfo) => Console.HideWarnings = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, Console.HideWarnings, true)
            });
#endif

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.consoleontop",
                ParseCommandLineArg = (string value) => Console.AlwaysOnTop = true,

                Config_Name = "AlwaysOnTop",
                Config_Category = Config_Category_Console,
                ParseConfigArg = (ArgInfo argInfo) => Console.AlwaysOnTop = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, Console.AlwaysOnTop, true)
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.consoledst",
                ParseCommandLineArg = (string value) => Console.SetTitleOnInit = false,

                Config_Name = "SetTitleOnInit",
                Config_Category = Config_Category_Console,
                ParseConfigArg = (ArgInfo argInfo) => Console.SetTitleOnInit = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, Console.SetTitleOnInit, true)
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.magenta",
                ParseCommandLineArg = (string value) => Console.DisplayMode = Console.DisplayModeEnum.MAGENTA
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.rainbow",
                ParseCommandLineArg = (string value) => Console.DisplayMode = Console.DisplayModeEnum.RAINBOW
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.randomrainbow",
                ParseCommandLineArg = (string value) => Console.DisplayMode = Console.DisplayModeEnum.RANDOMRAINBOW
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--lemonloader",
                ParseCommandLineArg = (string value) => Console.DisplayMode = Console.DisplayModeEnum.LEMON
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                Config_Name = "DisplayMode",
                Config_Category = Config_Category_Console,
                ParseConfigArg = (ArgInfo argInfo) =>
                {
                    int mode = iniFile.GetInt(argInfo.Config_Category, argInfo.Config_Name, (int)Console.DisplayMode, true);
                    if ((mode >= (int)Console.DisplayModeEnum.NORMAL)
                        && (mode <= (int)Console.DisplayModeEnum.LEMON))
                        Console.DisplayMode = (Console.DisplayModeEnum)mode;
                }
            });
            #endregion

            #region Core
            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.loadmodeplugins",
                ParseCommandLineArg = (string value) =>
                {
                    int mode = 0;
                    if (int.TryParse(value, out mode)
                        && (mode >= (int)Core.LoadModeEnum.NORMAL)
                        && (mode <= (int)Core.LoadModeEnum.BOTH))
                        Core.LoadMode_Plugins = (Core.LoadModeEnum)mode;
                },

                Config_Name = "LoadMode_Plugins",
                Config_Category = Config_Category_Core,
                ParseConfigArg = (ArgInfo argInfo) =>
                {
                    int mode = iniFile.GetInt(argInfo.Config_Category, argInfo.Config_Name, (int)Core.LoadMode_Plugins, true);
                    if ((mode >= (int)Core.LoadModeEnum.NORMAL)
                        && (mode <= (int)Core.LoadModeEnum.BOTH))
                        Core.LoadMode_Plugins = (Core.LoadModeEnum)mode;
                }
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.loadmodemods",
                ParseCommandLineArg = (string value) =>
                {
                    int mode = 0;
                    if (int.TryParse(value, out mode)
                        && (mode >= (int)Core.LoadModeEnum.NORMAL)
                        && (mode <= (int)Core.LoadModeEnum.BOTH))
                        Core.LoadMode_Mods = (Core.LoadModeEnum)mode;
                },

                Config_Name = "LoadMode_Mods",
                Config_Category = Config_Category_Core,
                ParseConfigArg = (ArgInfo argInfo) =>
                {
                    int mode = iniFile.GetInt(argInfo.Config_Category, argInfo.Config_Name, (int)Core.LoadMode_Mods, true);
                    if ((mode >= (int)Core.LoadModeEnum.NORMAL)
                        && (mode <= (int)Core.LoadModeEnum.BOTH))
                        Core.LoadMode_Mods = (Core.LoadModeEnum)mode;
                }
            });

#if !DEBUG
            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.debug",
                ParseCommandLineArg = (string value) => Core.DebugMode = true,

                Config_Name = "DebugMode",
                Config_Category = Config_Category_Core,
                ParseConfigArg = (ArgInfo argInfo) => Core.DebugMode = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, Core.DebugMode, true)
            });
#endif

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--quitfix",
                ParseCommandLineArg = (string value) => Core.QuitFix = true,

                Config_Name = "QuitFix",
                Config_Category = Config_Category_Core,
                ParseConfigArg = (ArgInfo argInfo) => Core.QuitFix = iniFile.GetBool(argInfo.Config_Category, argInfo.Config_Name, Core.QuitFix, true)
            });
            #endregion

            #region Logger
            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.maxlogs",
                ParseCommandLineArg = (string value) => { int val = Logger.Max_Logs; if (int.TryParse(value, out val)) Logger.Max_Logs = val; },

                Config_Name = "Max_Logs",
                Config_Category = Config_Category_Logger,
                ParseConfigArg = (ArgInfo argInfo) => Logger.Max_Logs = iniFile.GetInt(argInfo.Config_Category, argInfo.Config_Name, Logger.Max_Logs, true)
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.maxwarnings",
                ParseCommandLineArg = (string value) => { int val = Logger.Max_Warnings; if (int.TryParse(value, out val)) Logger.Max_Warnings = val; },

                Config_Name = "Max_Warnings",
                Config_Category = Config_Category_Logger,
                ParseConfigArg = (ArgInfo argInfo) => Logger.Max_Warnings = iniFile.GetInt(argInfo.Config_Category, argInfo.Config_Name, Logger.Max_Warnings, true)
            });

            ArgInfoTbl.Add(new ArgInfo
            {
                CommandLine = "--melonloader.maxerrors",
                ParseCommandLineArg = (string value) => { int val = Logger.Max_Errors; if (int.TryParse(value, out val)) Logger.Max_Errors = val; },
                
                Config_Name = "Max_Errors",
                Config_Category = Config_Category_Logger,
                ParseConfigArg = (ArgInfo argInfo) => Logger.Max_Errors = iniFile.GetInt(argInfo.Config_Category, argInfo.Config_Name, Logger.Max_Errors, true)
            });
            #endregion
        }

        internal static void Load()
        {
            ParseConfig();
            ParseCommandLine();
        }

        private static void ParseConfig()
        {
            if (ArgInfoTbl.Count <= 0)
                return;
            foreach (ArgInfo argInfo in ArgInfoTbl)
            {
                if ((argInfo.ParseConfigArg == null)
                    || string.IsNullOrEmpty(argInfo.Config_Category)
                    || string.IsNullOrEmpty(argInfo.Config_Name))
                    continue;
                argInfo.ParseConfigArg(argInfo);
            }
        }

        private static void ParseCommandLine()
        {
            if (ArgInfoTbl.Count <= 0)
                return;
            string[] args = Environment.GetCommandLineArgs();
            if ((args == null)
                || (args.Length <= 0))
                return;
            for (int i = 0; i < args.Length; i++)
            {
                string arg = args[i];
                if (string.IsNullOrEmpty(arg))
                    continue;
                ArgInfo argInfo = GetArgInfoFromCommandLine(arg);
                if ((argInfo == null)
                    || (argInfo.ParseCommandLineArg == null))
                    continue;
                string value = null;
                if ((i + 1) < args.Length)
                    value = args[i + 1];
                argInfo.ParseCommandLineArg(value);
            }
        }

        private static ArgInfo GetArgInfoFromCommandLine(string cmd)
        {
            foreach (ArgInfo argInfo in ArgInfoTbl)
            {
                if (!string.IsNullOrEmpty(argInfo.CommandLine)
                    && argInfo.CommandLine.Equals(cmd))
                    return argInfo;
            }
            return null;
        }

        #region Args
        public static class AnalyticsBlocker
        {
            static AnalyticsBlocker() => ShouldDAB = false;
            public static bool ShouldDAB { get; internal set; }
        }

        public static class AssemblyGenerator
        {
            static AssemblyGenerator()
            {
                ForceRegeneration = false;
                ForceVersion_Dumper = "0.0.0.0";
                ForceVersion_Il2CppAssemblyUnhollower = "0.0.0.0";
                ForceVersion_UnityDependencies = "0.0.0.0";
            }
            public static bool ForceRegeneration { get; internal set; }
            public static string ForceVersion_Dumper { get; internal set; }
            public static string ForceVersion_Il2CppAssemblyUnhollower { get; internal set; }
            public static string ForceVersion_UnityDependencies { get; internal set; }
        }

        public static class Console
        {
            static Console()
            {
                DisplayMode = DisplayModeEnum.NORMAL;
                Enabled = true;
                AlwaysOnTop = false;
                HideWarnings = false;
                SetTitleOnInit = true;
            }
            public enum DisplayModeEnum
            {
                NORMAL,
                MAGENTA,
                RAINBOW,
                RANDOMRAINBOW,
                LEMON
            }
            public static DisplayModeEnum DisplayMode { get; internal set; }
            public static bool Enabled { get; internal set; }
            public static bool AlwaysOnTop { get; internal set; }
            public static bool HideWarnings { get; internal set; }
            public static bool SetTitleOnInit { get; internal set; }
        }

        public static class Core
        {
            static Core()
            {
                LoadMode_Plugins = LoadModeEnum.NORMAL;
                LoadMode_Mods = LoadModeEnum.NORMAL;
                QuitFix = false;
#if DEBUG
                DebugMode = true;
#else
                DebugMode = false;
#endif
            }
            public enum LoadModeEnum
            {
                NORMAL,
                DEV,
                BOTH
            }
            public static LoadModeEnum LoadMode_Plugins { get; internal set; }
            public static LoadModeEnum LoadMode_Mods { get; internal set; }
            public static bool DebugMode { get; internal set; }
            public static bool QuitFix { get; internal set; }
        }

        public static class Logger
        {
            static Logger()
            {
                Max_Logs = 10;
                Max_Warnings = 100;
                Max_Errors = 100;
            }
            public static int Max_Logs { get; internal set; }
            public static int Max_Warnings { get; internal set; }
            public static int Max_Errors { get; internal set; }
        }
        #endregion
    }
}

File: Bootstrap/Utils/Console/CommandLine.cpp
#ifndef PORT_DISABLE
#include <Windows.h>
#include "CommandLine.h"
#include "Console.h"
#include "../Base/Core.h"
#include "Logger.h"
#include "Debug.h"
#include "AnalyticsBlocker.h"
#include "../Managers/InternalCalls.h"
#include "AssemblyGenerator.h"
#include "../Managers/Game.h"

int CommandLine::argc = NULL;
char* CommandLine::argv[64];
IniFile* CommandLine::iniFile = NULL;

void CommandLine::Read()
{
	ReadIniFile();
	char* nextchar = NULL;
	char* curchar = strtok_s(GetCommandLineA(), " ", &nextchar);
	while (curchar && (argc < 63))
	{
		argv[argc++] = curchar;
		curchar = strtok_s(0, " ", &nextchar);
	}
	argv[argc] = 0;
	if (argc <= 0)
		return;
	for (int i = 0; i < argc; i++)
	{
		const char* command = argv[i];
		if (command == NULL)
			continue;
		else if (strstr(command, "--quitfix") != NULL)
			Core::QuitFix = true;
		else if (strstr(command, "--melonloader.consoleontop") != NULL)
			Console::AlwaysOnTop = true;
		else if (strstr(command, "--melonloader.magenta") != NULL)
			Console::Mode = Console::DisplayMode::MAGENTA;
		else if (strstr(command, "--melonloader.rainbow") != NULL)
			Console::Mode = Console::DisplayMode::RAINBOW;
		else if (strstr(command, "--melonloader.randomrainbow") != NULL)
			Console::Mode = Console::DisplayMode::RANDOMRAINBOW;
		else if (strstr(command, "--melonloader.dab") != NULL)
			AnalyticsBlocker::ShouldDAB = true;
		else if (strstr(command, "--melonloader.loadmodeplugins") != NULL)
		{
			int loadmode = atoi(argv[i + 1]);
			if (loadmode < 0)
				loadmode = 0;
			else if (loadmode > 2)
				loadmode = 0;
			InternalCalls::MelonHandler::LoadModeForPlugins = (InternalCalls::MelonHandler::LoadMode)loadmode;
		}
		else if (strstr(command, "--melonloader.loadmodemods") != NULL)
		{
			int loadmode = atoi(argv[i + 1]);
			if (loadmode < 0)
				loadmode = 0;
			else if (loadmode > 2)
				loadmode = 0;
			InternalCalls::MelonHandler::LoadModeForMods = (InternalCalls::MelonHandler::LoadMode)loadmode;
		}
		else if (strstr(command, "--melonloader.agfvunity"))
		{
			std::string version = argv[i + 1];
			AssemblyGenerator::ForceVersion_UnityDependencies = new char[version.size() + 1];
			std::copy(version.begin(), version.end(), AssemblyGenerator::ForceVersion_UnityDependencies);
			AssemblyGenerator::ForceVersion_UnityDependencies[version.size()] = '\0';
		}
		else if (strstr(command, "--melonloader.agfvdumper"))
		{
			std::string version = argv[i + 1];
			AssemblyGenerator::ForceVersion_Il2CppDumper = new char[version.size() + 1];
			std::copy(version.begin(), version.end(), AssemblyGenerator::ForceVersion_Il2CppDumper);
			AssemblyGenerator::ForceVersion_Il2CppDumper[version.size()] = '\0';
		}
		else if (strstr(command, "--melonloader.agfvunhollower"))
		{
			std::string version = argv[i + 1];
			AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower = new char[version.size() + 1];
			std::copy(version.begin(), version.end(), AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower);
			AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower[version.size()] = '\0';
		}
#ifndef DEBUG
		else if (strstr(command, "--melonloader.debug") != NULL)
			Debug::Enabled = true;
		else if (strstr(command, "--melonloader.hideconsole") != NULL)
			Console::ShouldHide = true;
		else if (strstr(command, "--melonloader.hidewarnings") != NULL)
			Console::HideWarnings = true;
		else if (strstr(command, "--melonloader.maxlogs") != NULL)
			Logger::MaxLogs = GetIntFromConstChar(argv[i + 1], 10);
		else if (strstr(command, "--melonloader.maxwarnings") != NULL)
			Logger::MaxWarnings = GetIntFromConstChar(argv[i + 1], 10);
		else if (strstr(command, "--melonloader.maxerrors") != NULL)
			Logger::MaxErrors = GetIntFromConstChar(argv[i + 1], 10);
#endif
	}
}

int CommandLine::GetIntFromConstChar(const char* str, int defaultval)
{
	if ((str == NULL) || (*str == '\0'))
		return defaultval;
	bool negate = (str[0] == '-');
	if ((*str == '+') || (*str == '-'))
		++str;
	if (*str == '\0')
		return defaultval;
	int result = 0;
	while (*str)
	{
		if ((*str >= '0') && (*str <= '9'))
			result = ((result * 10) - (*str - '0'));
		else
			return defaultval;
		++str;
	}
	return (negate ? result : -result);
}

void CommandLine::ReadIniFile()
{
	if (iniFile == NULL)
		iniFile = new IniFile((std::string(Game::BasePath) + "\\MelonLoader\\LaunchOptions.ini"));

#ifndef DEBUG
	Debug::Enabled = (!iniFile->ReadValue("Core", "Debug").empty() && iniFile->ReadValue("Core", "Debug")._Equal("true"));
	iniFile->WriteValue("Core", "Debug", (Debug::Enabled ? "true" : "false"));
#endif
	Core::QuitFix = (!iniFile->ReadValue("Core", "QuitFix").empty() && iniFile->ReadValue("Core", "QuitFix")._Equal("true"));
	iniFile->WriteValue("Core", "QuitFix", (Core::QuitFix ? "true" : "false"));

#ifndef DEBUG
	Console::ShouldHide = (!iniFile->ReadValue("Console", "Enabled").empty() && iniFile->ReadValue("Console", "Enabled")._Equal("false"));
	iniFile->WriteValue("Console", "Enabled", (Console::ShouldHide ? "false" : "true"));
#endif
	Console::Mode = (iniFile->ReadValue("Console", "Mode").empty() ? Console::Mode : (
		iniFile->ReadValue("Console", "Mode")._Equal("1") ? Console::DisplayMode::MAGENTA : (
			iniFile->ReadValue("Console", "Mode")._Equal("2") ? Console::DisplayMode::RAINBOW :
			(iniFile->ReadValue("Console", "Mode")._Equal("3") ? Console::DisplayMode::RANDOMRAINBOW : Console::Mode))));
	iniFile->WriteValue("Console", "Mode", std::to_string(Console::Mode));
	Console::AlwaysOnTop = (!iniFile->ReadValue("Console", "AlwaysOnTop").empty() && iniFile->ReadValue("Console", "AlwaysOnTop")._Equal("true"));
	iniFile->WriteValue("Console", "AlwaysOnTop", (Console::AlwaysOnTop ? "true" : "false"));
#ifndef DEBUG
	Console::HideWarnings = (!iniFile->ReadValue("Console", "HideWarnings").empty() && iniFile->ReadValue("Console", "HideWarnings")._Equal("true"));
	iniFile->WriteValue("Console", "HideWarnings", (Console::AlwaysOnTop ? "true" : "false"));

	Logger::MaxLogs = (!iniFile->ReadValue("Logger", "MaxLogs").empty() ? GetIntFromConstChar(iniFile->ReadValue("Console", "MaxLogs").c_str(), Logger::MaxLogs) : Logger::MaxLogs);
	iniFile->WriteValue("Logger", "MaxLogs", std::to_string(Logger::MaxLogs));
	Logger::MaxWarnings = (!iniFile->ReadValue("Logger", "MaxWarnings").empty() ? GetIntFromConstChar(iniFile->ReadValue("Console", "MaxWarnings").c_str(), Logger::MaxWarnings) : Logger::MaxWarnings);
	iniFile->WriteValue("Logger", "MaxWarnings", std::to_string(Logger::MaxWarnings));
	Logger::MaxErrors = (!iniFile->ReadValue("Logger", "MaxErrors").empty() ? GetIntFromConstChar(iniFile->ReadValue("Console", "MaxErrors").c_str(), Logger::MaxErrors) : Logger::MaxErrors);
	iniFile->WriteValue("Logger", "MaxErrors", std::to_string(Logger::MaxErrors));
#endif

	AnalyticsBlocker::ShouldDAB = (!iniFile->ReadValue("AnalyticsBlocker", "ShouldDAB").empty() && iniFile->ReadValue("AnalyticsBlocker", "ShouldDAB")._Equal("true"));
	iniFile->WriteValue("AnalyticsBlocker", "ShouldDAB", (AnalyticsBlocker::ShouldDAB ? "true" : "false"));

	InternalCalls::MelonHandler::LoadModeForPlugins = (iniFile->ReadValue("LoadMode", "Plugins").empty() ? InternalCalls::MelonHandler::LoadMode::NORMAL : (
		iniFile->ReadValue("LoadMode", "Plugins")._Equal("1") ? InternalCalls::MelonHandler::LoadMode::DEV : (
			iniFile->ReadValue("LoadMode", "Plugins")._Equal("2") ? InternalCalls::MelonHandler::LoadMode::BOTH : InternalCalls::MelonHandler::LoadMode::NORMAL)));
	iniFile->WriteValue("LoadMode", "Plugins", std::to_string(InternalCalls::MelonHandler::LoadModeForPlugins));
	InternalCalls::MelonHandler::LoadModeForMods = (iniFile->ReadValue("LoadMode", "Mods").empty() ? InternalCalls::MelonHandler::LoadMode::NORMAL : (
		iniFile->ReadValue("LoadMode", "Mods")._Equal("1") ? InternalCalls::MelonHandler::LoadMode::DEV : (
			iniFile->ReadValue("LoadMode", "Mods")._Equal("2") ? InternalCalls::MelonHandler::LoadMode::BOTH : InternalCalls::MelonHandler::LoadMode::NORMAL)));
	iniFile->WriteValue("LoadMode", "Mods", std::to_string(InternalCalls::MelonHandler::LoadModeForMods));

	if (iniFile->ReadValue("AssemblyGenerator", "ForceUnityDependencies").empty() || !iniFile->ReadValue("AssemblyGenerator", "ForceUnityDependencies")._Equal("true"))
	{
		iniFile->WriteValue("AssemblyGenerator", "ForceUnityDependencies", "false");
		iniFile->WriteValue("AssemblyGenerator", "ForceUnityDependencies_Version", "0.0.0.0");
	}
	else
	{
		std::string version = iniFile->ReadValue("AssemblyGenerator", "ForceUnityDependencies_Version");
		if (!version.empty())
		{
			AssemblyGenerator::ForceVersion_UnityDependencies = new char[version.size() + 1];
			std::copy(version.begin(), version.end(), AssemblyGenerator::ForceVersion_UnityDependencies);
			AssemblyGenerator::ForceVersion_UnityDependencies[version.size()] = '\0';
		}
		iniFile->WriteValue("AssemblyGenerator", "ForceUnityDependencies_Version", (std::string(AssemblyGenerator::ForceVersion_UnityDependencies).empty() ? "0.0.0.0" : AssemblyGenerator::ForceVersion_UnityDependencies));
	}
	if (iniFile->ReadValue("AssemblyGenerator", "ForceIl2CppDumper").empty() || !iniFile->ReadValue("AssemblyGenerator", "ForceIl2CppDumper")._Equal("true"))
	{
		iniFile->WriteValue("AssemblyGenerator", "ForceIl2CppDumper", "false");
		iniFile->WriteValue("AssemblyGenerator", "ForceIl2CppDumper_Version", "0.0.0.0");
	}
	else
	{
		std::string version = iniFile->ReadValue("AssemblyGenerator", "ForceIl2CppDumper_Version");
		if (!version.empty())
		{
			AssemblyGenerator::ForceVersion_Il2CppDumper = new char[version.size() + 1];
			std::copy(version.begin(), version.end(), AssemblyGenerator::ForceVersion_Il2CppDumper);
			AssemblyGenerator::ForceVersion_Il2CppDumper[version.size()] = '\0';
		}
		iniFile->WriteValue("AssemblyGenerator", "ForceIl2CppDumper_Version", (std::string(AssemblyGenerator::ForceVersion_Il2CppDumper).empty() ? "0.0.0.0" : AssemblyGenerator::ForceVersion_Il2CppDumper));
	}
	if (iniFile->ReadValue("AssemblyGenerator", "ForceIl2CppAssemblyUnhollower").empty() || !iniFile->ReadValue("AssemblyGenerator", "ForceIl2CppAssemblyUnhollower")._Equal("true"))
	{
		iniFile->WriteValue("AssemblyGenerator", "ForceIl2CppAssemblyUnhollower", "false");
		iniFile->WriteValue("AssemblyGenerator", "ForceIl2CppAssemblyUnhollower_Version", "0.0.0.0");
	}
	else
	{
		std::string version = iniFile->ReadValue("AssemblyGenerator", "ForceIl2CppAssemblyUnhollower_Version");
		if (!version.empty())
		{
			AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower = new char[version.size() + 1];
			std::copy(version.begin(), version.end(), AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower);
			AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower[version.size()] = '\0';
		}
		iniFile->WriteValue("AssemblyGenerator", "ForceIl2CppAssemblyUnhollower_Version", (std::string(AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower).empty() ? "0.0.0.0" : AssemblyGenerator::ForceVersion_Il2CppAssemblyUnhollower));
	}
}
#endif